### PR TITLE
test: Add tests for Hub refreshPremiumDebt

### LINKED ISSUE #314: Spoke Tests: Refresh

### PR DESCRIPTION
**Pull request by miguelmtzinf**
Closes #314

### CODE REVIEW

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+      );
+  }
+
+  function _assertPremiumData(PremiumDataLocal memory a, PremiumDataLocal memory b) internal pure {
+    assertEq(a.premiumDebtShares, b.premiumDebtShares, 'premium debt shares do not match');
+    assertEq(a.premiumOffset, b.premiumOffset, 'premium offset do not match');
+    assertEq(a.realizedPremium, b.realizedPremium, 'realized premium do not match');
+  }
+
+  // only assets listed can be refreshed
```
Comment (yan-man) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:48]: saw before you were using /// rather than // for test descriptions, do you care either way? just want to keep my stuff consistent with whatever we're doing 

https://github.com/aave/aave-v4/commit/767d002973982a1613c52d2a8533277e7505b5c2#diff-c3b728636ae9e535fdd38fa2792f03669a2ac8dfa2213cf1fb7d16d233141da7

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        uint256(newPremiumOffset + int256(initPremiumOffset)),
+        uint256(newRealizedPremium + int256(initRealizedPremium))
+      );
+      _assertPremiumData(_loadAssetPremiumData(assetId), newData);
+      _assertPremiumData(_loadSpokePremiumData(assetId, address(spoke1)), newData);
+    }
+  }
+
+  // refresh premium data of DAI by 3 different spokes
+  /// forge-config: default.fuzz.runs = 100
```
Comment (yan-man) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:225]: why restrict fuzz runs?

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+  // premium data values cannot go negative
+  function test_refresh_from_init_revertsWith_underflow() public {
+    uint256 assetId = 0;
+    _assertPremiumData(_loadAssetPremiumData(assetId), PremiumDataLocal(0, 0, 0));
+    _assertPremiumData(_loadSpokePremiumData(assetId, address(spoke1)), PremiumDataLocal(0, 0, 0));
+
+    vm.expectRevert(stdError.arithmeticError);
+    vm.prank(address(spoke1));
+    hub.refreshPremiumDebt(assetId, -1, -1, -1);
+  }
```
Comment (yan-man) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:95]: worth it to add another test to fuzz each of the refreshPremiumDebt params negative input value separately?

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    uint256 targetToChange,
+    uint256 newValue
+  ) public {
+    uint256 assetId = daiAssetId;
+    address[3] memory spokes = [address(spoke1), address(spoke2), address(spoke3)];
+    spokeIndex = bound(spokeIndex, 0, spokes.length - 1);
+    initPremiumShares = bound(initPremiumShares, 0, MAX_SUPPLY_AMOUNT / 2);
+    initPremiumOffset = bound(initPremiumOffset, 0, MAX_SUPPLY_AMOUNT / 2);
+    initRealizedPremium = bound(initRealizedPremium, 0, MAX_SUPPLY_AMOUNT / 2);
+    targetToChange = bound(targetToChange, 0, 3); // [premiumDebtShares, premiumOffset, realizedPremium]
```
Comment (yan-man) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:240]: I think can just bound to 2?
```suggestion
    targetToChange = bound(targetToChange, 0, 2); // [premiumDebtShares, premiumOffset, realizedPremium]
```

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    uint256 notListedAsset = hub.assetCount() + 1;
+
+    vm.expectRevert(ILiquidityHub.AssetNotListed.selector);
+    vm.prank(address(spoke1));
+    hub.refreshPremiumDebt(notListedAsset, 1, 1, 1);
+  }
+
+  // only authorized spokes can refresh
+  function test_refresh_revertsWith_AuthorizedSpoke() public {
+    // TODO
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:59]:
```suggestion
    vm.skip(true, 'todo after spoke auth');
```

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    DataTypes.SpokeData memory spokeData = hub.getSpoke(assetId, spoke);
+    return
+      PremiumDataLocal(
+        spokeData.premiumDrawnShares,
+        spokeData.premiumOffset,
+        spokeData.realizedPremium
+      );
+  }
+
+  function _assertPremiumData(PremiumDataLocal memory a, PremiumDataLocal memory b) internal pure {
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:42]: i'd just name this as an `assertEq` tbh

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    uint256 assetId = 0;
+    _assertPremiumData(_loadAssetPremiumData(assetId), PremiumDataLocal(0, 0, 0));
+    _assertPremiumData(_loadSpokePremiumData(assetId, address(spoke1)), PremiumDataLocal(0, 0, 0));
+
+    vm.expectRevert(stdError.arithmeticError);
+    vm.prank(address(spoke1));
+    hub.refreshPremiumDebt(assetId, -1, -1, -1);
+  }
+
+  // refresh premium data with positive values from initial state
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:97]: currrently it's not positive values when the latter three fuzzed values are zero but nbd

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+  function test_refresh_from_init_fuzz_positive_amounts(
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:98]: we'd be adding more assertions in the refresh method that total debt has not changed 
ideally it can only decrease on a refresh and remain the same on all other cases

refresh can **never** increase total debt, i would change this method to negative amounts for now tbh

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    assetId = bound(assetId, 0, hub.assetCount() - 1);
+    initPremiumShares = bound(initPremiumShares, 0, MAX_SUPPLY_AMOUNT / 2);
+    initPremiumOffset = bound(initPremiumOffset, 0, MAX_SUPPLY_AMOUNT / 2);
+    initRealizedPremium = bound(initRealizedPremium, 0, MAX_SUPPLY_AMOUNT / 2);
+    premiumDebtShares = bound(premiumDebtShares, 0, MAX_SUPPLY_AMOUNT / 2);
+    premiumOffset = bound(premiumOffset, 0, MAX_SUPPLY_AMOUNT / 2);
+    realizedPremium = bound(realizedPremium, 0, MAX_SUPPLY_AMOUNT / 2);
+
+    // random sign: negative if even
+    int256 newPremiumShares = (initPremiumShares + premiumDebtShares) % 2 == 0
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:157]:
```suggestion
    // random sign: negative if even
    int256 newPremiumShares = (vm.randomUint()) % 2 == 0
```

or we add a `_randomChance(5)` helper in baseTest

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+      int256(15)
+    );
+    vm.prank(address(spoke1));
+    hub.refreshPremiumDebt(assetId, 5, 10, 15);
+
+    _assertPremiumData(_loadAssetPremiumData(assetId), PremiumDataLocal(5, 10, 15));
+    _assertPremiumData(
+      _loadSpokePremiumData(assetId, address(spoke1)),
+      PremiumDataLocal(5, 10, 15)
+    );
```
Comment (DhairyaSethi) [inline tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:83]: we can also assert the correct total debt increase (and that ONLY premium debt has increased)

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 2 above)
Comment (miguelmtzinf) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:225]: takes a long time otherwise. I can remove and decrease iterations maybe, or just leave it as it is and create a new profile for the CI where we used reduced number of fuzz runs?

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 3 above)
Comment (miguelmtzinf) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:95]: added no-fuzz test `test_refresh_from_init_revertsWith_underflow_separately`. lmk if works or you think fuzz adds more value here

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 4 above)
Comment (miguelmtzinf) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:240]: good catch! fixed

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 2 above)
Comment (DhairyaSethi) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:225]: i'm ok w overriding fuzz runs on the default profile (local) to 256 or 1000 runs at the very least. but ci should be running the 10k runs 
it's parallelised on the ci, if it takes over 20 mins we can reconsider for particular tests

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 7 above)
Comment (miguelmtzinf) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:97]: i mean going from init (zero) and using positive inputs/delta. Starting from random positive state in other tests (e.g. `test_refresh_fuzz_amounts`)

File: tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol (hunk 8 above)
Comment (miguelmtzinf) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:98]: this test goes from 0 to >0, so cannot add negative delta here. 
`test_refresh_fuzz_amounts` does the work you suggest I think. Is it fine to you or should I add something else?

Comment (DhairyaSethi) [reply tests/unit/LiquidityHub/LiquidityHub.RefreshPremiumDebt.t.sol:98]: yeah my point was we'd need to rm this test once we add the assertions on refresh that total debt cannot increase but yeah since the other test covers both (including -ve) cases it's fine

**Comment by DhairyaSethi** [comment]
superseded by #597
